# Week 7: Task 4 - Iris Flower Classification
## Exploratory Data Analysis (Monday Deliverable)

**Internship:** Arch Technologies, Machine Learning Domain, Month 2  
**Task:** Task 4 — Iris Flower Classification (First Half)  
**Author:** Sharjeel Shahzad  

### Overview & Objectives
The goal of this project is to build an end-to-end Machine Learning pipeline to classify iris flowers into three distinct species (**Iris-setosa**, **Iris-versicolor**, and **Iris-virginica**) based on four morphological measurements:
1. `SepalLengthCm`: Sepal length in centimeters
2. `SepalWidthCm`: Sepal width in centimeters
3. `PetalLengthCm`: Petal length in centimeters
4. `PetalWidthCm`: Petal width in centimeters

In this initial exploration notebook (**Monday deliverable**), we perform:
- Loading the Kaggle UCI Iris dataset from `data/iris-dataset.csv`
- Dropping the extraneous `Id` row identifier column
- Analyzing structural properties: `.shape`, `.dtypes`, `.info()`
- Computing statistical summaries: `.describe()`
- Verifying missing values and inspecting class distribution and balance


## 1. Environment Setup & Data Loading
We import core analytical libraries and resolve file paths dynamically to support execution whether run from the `week-7/notebooks/` directory or root.


In [1]:
from pathlib import Path
import pandas as pd
import numpy as np

# Configure path dynamically to dataset
data_path = Path("../data/iris-dataset.csv")
if not data_path.exists():
    data_path = Path("week-7/data/iris-dataset.csv")

# Load raw dataset into a Pandas DataFrame
df = pd.read_csv(data_path)
print(f"Dataset successfully loaded from: {data_path}")
print(f"Raw shape: {df.shape}")
df.head()


Dataset successfully loaded from: ../data/iris-dataset.csv
Raw shape: (150, 6)


   Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm      Species
0   1            5.1           3.5            1.4           0.2  Iris-setosa
1   2            4.9           3.0            1.4           0.2  Iris-setosa
2   3            4.7           3.2            1.3           0.2  Iris-setosa
3   4            4.6           3.1            1.5           0.2  Iris-setosa
4   5            5.0           3.6            1.4           0.2  Iris-setosa


## 2. Data Cleaning: Dropping the `Id` Column
The `Id` column is an arbitrary row index assigned in the Kaggle UCI dataset. It carries no biological signal and would introduce spurious correlation if retained. We drop it immediately.


In [2]:
# Drop non-informative Id column
if "Id" in df.columns:
    df.drop(columns=["Id"], inplace=True)
    print("Column 'Id' dropped successfully.")

print(f"New DataFrame dimensions: {df.shape}")
df.head()


Column 'Id' dropped successfully.
New DataFrame dimensions: (150, 5)


   SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm      Species
0            5.1           3.5            1.4           0.2  Iris-setosa
1            4.9           3.0            1.4           0.2  Iris-setosa
2            4.7           3.2            1.3           0.2  Iris-setosa
3            4.6           3.1            1.5           0.2  Iris-setosa
4            5.0           3.6            1.4           0.2  Iris-setosa


## 3. Structural Properties & Data Types
We inspect dimensions, data types, and non-null counts using `.shape`, `.dtypes`, and `.info()`.


In [3]:
print("--- Dataset Shape ---")
print(f"Total Rows: {df.shape[0]}, Total Columns: {df.shape[1]}
")

print("--- Column Data Types ---")
print(df.dtypes)
print("
--- Concise DataFrame Summary ---")
df.info()


--- Dataset Shape ---
Total Rows: 150, Total Columns: 5

--- Column Data Types ---
SepalLengthCm    float64
SepalWidthCm     float64
PetalLengthCm    float64
PetalWidthCm     float64
Species           object
dtype: object

--- Concise DataFrame Summary ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   SepalLengthCm  150 non-null    float64
 1   SepalWidthCm   150 non-null    float64
 2   PetalLengthCm  150 non-null    float64
 3   PetalWidthCm   150 non-null    float64
 4   Species        150 non-null    object 
dtypes: float64(4), object(1)
memory usage: 6.0+ KB


## 4. Descriptive Summary Statistics
We compute summary metrics (mean, standard deviation, percentiles, minimum, and maximum) for all four numerical features.


In [4]:
# Summary statistics for numerical features
df.describe()


       SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm
count     150.000000    150.000000     150.000000    150.000000
mean        5.843333      3.054000       3.758667      1.198667
std         0.828066      0.433594       1.764420      0.763161
min         4.300000      2.000000       1.000000      0.100000
25%         5.100000      2.800000       1.600000      0.300000
50%         5.800000      3.000000       4.350000      1.300000
75%         6.400000      3.300000       5.100000      1.800000
max         7.900000      4.400000       6.900000      2.500000


### Per-Species Mean & Standard Deviation
Breaking down feature means by species provides initial quantitative evidence of species divergence.


In [5]:
# Per-species feature means
df.groupby("Species").mean()


                 SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm
Species                                                                  
Iris-setosa              5.006         3.418          1.464         0.244
Iris-versicolor          5.936         2.770          4.260         1.326
Iris-virginica           6.588         2.974          5.552         2.026


## 5. Missing Value & Data Integrity Verification
We check for null, missing, or NaN values across every feature.


In [6]:
# Check for missing values
missing_counts = df.isnull().sum()
print("Missing values per column:")
print(missing_counts)
print(f"Total missing values in entire dataset: {missing_counts.sum()}")


Missing values per column:
SepalLengthCm    0
SepalWidthCm     0
PetalLengthCm    0
PetalWidthCm     0
Species          0
dtype: int64
Total missing values in entire dataset: 0


## 6. Class Distribution & Balance Verification
We verify the sample count and frequency distribution for each of the three target species.


In [7]:
# Value counts and percentages per species
species_counts = df["Species"].value_counts()
species_pct = df["Species"].value_counts(normalize=True) * 100

distribution_df = pd.DataFrame({
    "Sample Count": species_counts,
    "Percentage (%)": species_pct
})
print("Class Distribution:")
distribution_df


Class Distribution:


                 Sample Count  Percentage (%)
Species                                      
Iris-setosa                50       33.333333
Iris-versicolor            50       33.333333
Iris-virginica             50       33.333333


## 7. Monday Exploration Summary & Findings

| Property | Value | Observation |
| :--- | :--- | :--- |
| **Total Rows** | 150 | Exactly 50 observations per class |
| **Total Features** | 4 | Continuous measurements in centimeters (`float64`) |
| **Target Variable** | `Species` | Categorical string with 3 distinct classes |
| **Missing Values** | 0 | Clean, complete dataset requiring no imputation |
| **Class Balance** | 33.33% / 33.33% / 33.33% | Perfectly balanced across all three species |

### Key Takeaways:
1. **Biological Divergence in Petals**: `Iris-setosa` exhibits markedly smaller petals (`mean PetalLength = 1.46 cm`, `PetalWidth = 0.24 cm`) compared to `Iris-versicolor` (`4.26 cm` and `1.33 cm`) and `Iris-virginica` (`5.55 cm` and `2.03 cm`).
2. **Sepal Overlap**: `SepalLength` and `SepalWidth` display greater overlap across species, suggesting petal dimensions will be our primary discriminative features.
3. **Next Steps (Tuesday)**: Perform comprehensive multi-dimensional data visualization, pair plots, and scatter plots to analyze class boundaries.
